server/Arcadia.Assistant.Notifications.Push/PushNotificationsActor.cs

```diff
@@ -0,0 +1,141 @@
+                            this.Self,
+                            success: x => new SendPushNotificationResult(x),
+                            failure: err => new SendPushNotificationResultFinish(err.Message));
+                    break;
+
+                case SendPushNotificationResult msg:
+                    this.HandleSendPushNotificationResults(msg.Responses)
+                        .PipeTo(
+                            this.Self,
+                            success: () => SendPushNotificationResultFinish.Instance,
+                            failure: err => new SendPushNotificationResultFinish(err.Message));
+                    break;
+
+                case SendPushNotificationResultFinish msg:
+                    if (msg.ErrorMessage != null)
+                    {
+                        this.logger.Warning(msg.ErrorMessage);
+                    }
+
+                    break;
+            }
+        }
+
+        private Task<HttpResponseMessage[]> SendPushNotification(PushNotification message)
+        {
+            this.logger.Debug("Push notification message received");
+
+            var jsonMessage = this.SerializeNotification(message);
+            this.logger.Debug($"Serialized notification message: {jsonMessage}");
+
+            var requests = this.pushSettings.ApplicationPushUrls
+                .Select(url => this.SendPushNotificationRequest(url, jsonMessage));
+
+            return Task.WhenAll(requests);
+        }
+
+        private async Task HandleSendPushNotificationResults(IEnumerable<HttpResponseMessage> responses)
+        {
+            foreach (var response in responses)
+            {
+                if (response.StatusCode == HttpStatusCode.Accepted)
+                {
+                    this.logger.Debug($"Push notification was succesfully sent to {response.RequestMessage.RequestUri}");
+                }
+                else
+                {
+                    var content = await response.Content.ReadAsStringAsync();
+                    this.logger.Warning($"Push notification to {response.RequestMessage.RequestUri} failed: {content}");
+                }
+            }
+        }
+
+        private Task<HttpResponseMessage> SendPushNotificationRequest(
+            string applicationPushUrl,
+            string pushNotificationContent)
+        {
+            var content = new StringContent(pushNotificationContent);
+            content.Headers.ContentType.MediaType = "application/json";
+
+            var request = new HttpRequestMessage(HttpMethod.Post, applicationPushUrl)
+            {
+                Content = content
+            };
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            content.Headers.Add("X-API-Token", this.pushSettings.ApiToken);
+
+            using (var httpClient = this.httpClientFactory.CreateClient())
+            {
+                return httpClient.SendAsync(request);
```

Comment:
i wonder how it works if Http Client might get disposed BEFORE SendAsync finished its job. I assume you just get lucky here because of some pooling. Better to await it here.

Reply:
Fixed.